### ISSUE TITLE
User Login via single sign-on

### ISSUE THREAD

**Issue by b-fein** [ls1intum/Artemis#2377]
### Is your feature request related to a problem?
All students already have an university account. I would like them to be able to use this account to login into Artemis so that no additional account-management with forgotten passwords/registration emails/etc. is necessary.

### Describe the solution you'd like
A single sign-on solution like Shibboleth would enable such a login.
As the Artemis instance used by the Austrian universities linked in the Readme (https://artemis.codeability.uibk.ac.at/#/) seems to have this feature already, this would probably be the easiest solution to implement.

### Describe alternatives you've considered
A full login via LDAP could solve this issue as well. Currently only additional information for users created in Artemis itself can be retrieved from LDAP.

**Comment by krusche** [comment]
Please note that the Shibboleth login is limited for single sign on. You would then need to make sure that the version control server (and potentially the continuous integration server) can also use Shibboleth (which does not seem to be easy) and you would need to find a way to sync the user rights between the systems.

LDAP seems to be the far better solution, when single sign on should be possible. It would need to be connected to both, Artemis and the version control server (and potentially also the continuous integration server ).

Then, you would need to make sure that the user is available in the version control server before the user rights are synced. One open issue would also be how to handle user groups.

**Comment by krusche** [comment]
As far as I know, https://artemis.codeability.uibk.ac.at/#/ only uses Shibboleth for Artemis.
Students can have a different password on Gitlab, but need to have the same user name.
I am not sure how the synchronization of groups (for tutors and instructors) would work then.

**Comment by LDAP** [comment]
I am currently investigating the necessary changes to enable Shibboleth (and every other SAML 2 IdP) as AuthenticationProvider in Artemis. 

- GitLab does natively support SAML 2.0 (Shibboleth is based on SAML)
- Jenkins does have a [SAML Plugin](https://plugins.jenkins.io/saml/). This needs to be evaluated. (low priority)
- Spring Boot does have [SAML 2.0 Support](https://docs.spring.io/spring-security/site/docs/5.2.1.RELEASE/reference/htmlsingle/#saml2) build in.

That means SAML 2 could be used to register users in Artemis. Artemis then creates the users in GitLab. If a student logs into GitLab the user is matched by mail-address which should be unique in an IdP (This feature is natively supported by GitLab). Only special case: When a user first logs into GitLab the user gets created and blocked by GitLab and Artemis fails to create the user. In this case Artemis should instead unblock the user.

I tested creating a user in Artemis and then login in GitLab via SAML - works like a charm.

**Comment by daniels98it** [comment]
We include our experience here, explaining our current setup enabling SSO authentication for ArTEMiS.

The original documentation is available here: [https://sharing-codeability.uibk.ac.at/sharing/artemis-user-documentation/-/wikis/development/Shibboleth-EduId-integration-with-ArTEMiS](https://sharing-codeability.uibk.ac.at/sharing/artemis-user-documentation/-/wikis/development/Shibboleth-EduId-integration-with-ArTEMiS) but requires registration to be accessed.

### Shibboleth - EduId integration

EduID is a wide-spread federated identity service for academic institutions. It is based on the [Shibboleth protocol](https://de.wikipedia.org/wiki/Shibboleth).

You can find organisational and technical information about EduId e.g. [here](http://eduid.at) or via the IT services of your organization.

Here we give an overview how to integrate ArTEMiS with EduId.

Involved Components:
- ArTEMiS instance
- Apache2
- Shibboleth SP
- Embedded Discovery Service (EDS) page (\*optional\*)

##### Shibboleth configuration

We won't cover in deep the Shibboleth set up in this Wiki page, but focus on the integration with Apache and ArTEMiS

[Example of Shibboleth setup on Linux with Apache](https://wiki.cac.washington.edu/display/infra/Install+Shibboleth+Service+Provider+on+Linux+and+Apache)

After a working Shibboleth Service Provider (SP) has been set up, using a single IDP or (\*optional\*) Discovery Service, we have to make sure that in our main Shibboleth config (ex. shibboleth2.xml) `ApplicationDefaults REMOTE_USER` contains all the needed information (in our case `"eppn subject-id pairwise-id persistent-id targeted-id"`). 

Additionally we added `ApplicationDefaults attributePrefix="AJP_"` to avoid conflicts.

##### Apache2

In our case, Shibboleth is used along with Apache2, managing a reverse proxy to our ArTEMiS instance.
 
For the Shibboleth integration in Apache2 we have (within the VirtualHost containing ArTEMiS) 2 **Location directives** for the subpaths `/Shibboleth.sso` and `/api/authenticate-shib` configured as follows:
```
    <Location >
        AuthType shibboleth
        ShibRequestSetting requireSession 1
        ShibUseHeaders On
        Require valid-user
    </Location>
```

**/Shibboleth.sso**: A secured Shibboleth path to start authentication process. We will use this URL to redirect the user to the Shibboleth login with `https://<artemis_url>/Shibboleth.sso/Login?target=https://<artemis_url>`

**/api/authenticate-shib**: The new ArTEMiS endpoint to handle Shibboleth logins. This endpoint has to be secured too, in order to ensure that the Shibboleth attributes (containing the data of the user) are included in the request.


##### Code Changes

In ArTEMiS we added an additional Endpoint "/authenticate-shib" to allow the login using the attributes provided by the shibboleth headers.

The main changes in ArTEMiS code occur in `UserJWTController.java`, were the additional endpoint ("/authenticate-shib") is implemented. Here we can access the user's attributes using `@RequestHeader HttpHeaders headers` which can then be used to implement a new **authorize** method.

Other important changes occur in `home.component.ts` where we have to take care of redirecting the user to the Shibboleth secured paths and call the new endpoint for login.


##### Login process diagramm

https://ibb.co/0M2Xz79

**Comment by 1-alex98** [comment]
Interesting our university also uses Shibboleth...

**Comment by dfuchss** [comment]
JFYI: We (at the KIT) plan to change the state of our PR #3002 to "Ready for Review" within this week. On our test servers, the Shibboleth login / registration feature works.

**Comment by dfuchss** [comment]
@b-fein @1-alex98 our feature is merged to develop. Maybe you could check whether this feature is applicable for you.

**Comment by dfuchss** [comment]
@daniels98it you might also check whether this feature is applicable for you; but I think your setting was a bit more complex (IdP pre selection) .. but our feature should support multiple IdPs  :)

**Comment by 1-alex98** [comment]
> @b-fein @1-alex98 our feature is merged to develop. Maybe you could check whether this feature is applicable for you.

Next semester Artemis will not be used. So this is a future project :D

**Comment by daniels98it** [comment]
> @daniels98it you might also check whether this feature is applicable for you; but I think your setting was a bit more complex (IdP pre selection) .. but our feature should support multiple IdPs :)

Thanks @dfuchss :slightly_smiling_face: 
I'll give it a look in the next days :+1:

**Comment by dfuchss** [comment]
Is this issue ready for close? Since SAML2 has been implemented, Single-Sign-On is operational :)

**Comment by b-fein** [comment]
We currently don’t have any courses using Artemis this semester, so I could not yet test it in production. As we use Shibboleth as well and it works for you, I’m confident the feature should work for us, too. Therefore, this issue can indeed be closed.

Thanks for implementing it!
